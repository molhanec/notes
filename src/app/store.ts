import { Action, configureStore, getDefaultMiddleware, ThunkAction } from '@reduxjs/toolkit'
import { connectRouter, routerMiddleware } from 'connected-react-router'
import { createBrowserHistory } from 'history'

import counterReducer from '../features/counter/counterSlice'
import notesReducer from '../features/notes/notesSlice'

export const history = createBrowserHistory()

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    router: connectRouter(history) as any,
    notes: notesReducer,
  },
  middleware: [...getDefaultMiddleware(), routerMiddleware(history)],
})

export type RootState = ReturnType<typeof store.getState>
export type AppThunk = ThunkAction<void, RootState, unknown, Action<string>>
