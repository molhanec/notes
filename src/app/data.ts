import { AppThunk } from "./store"
import { loadNotesAsync } from "../features/noteList/notesSlice"
import { setProgress, setSuccess, setError } from "../features/info/infoSlice"

const rootDataUrl =
  "https://private-anon-3ab0da907c-note10.apiary-mock.com/notes"

const checkResponseStatus = (response: Response) => {
  if (response.ok) {
    return response
  } else {
    return Promise.reject(
      `Error fetching data: ${response.status} ${response.statusText}`
    )
  }
}

export const fetchJson = <T>(path: string) =>
  fetch(rootDataUrl + path)
    .then(checkResponseStatus)
    .then(response => response!.json() as Promise<T>)

export const addNoteAsync = (title: string): AppThunk => async dispatch => {
  dispatch(setProgress())
  try {
    fetch(rootDataUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        title,
      }),
    }).then(checkResponseStatus)

    dispatch(setSuccess("New note added"))
    dispatch(loadNotesAsync())
  } catch {
    dispatch(setError("Cannot add new note"))
  }
}
