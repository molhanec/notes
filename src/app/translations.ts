import { useSelector } from "react-redux"
import { RootState } from "./store"

export type LanguageCodes = "cz" | "en"
export const defaultLanguage = "cz"

interface Phrases {
  appTitle: string
  noteLoadFailure: string
  notesLoadFailure: string

  noteCreateSuccess: string
  noteCreateFailure: string
  noteUpdateSuccess: string
  noteUpdateFailure: string
  noteDeleteSuccess: string
  noteDeleteFailure: string

  formTitle: string
  formSave: string
  formDelete: string

  editFormHeader: string
  newFormHeader: string
}

export type PhraseKeys = keyof Phrases

const enPhrases: Phrases = {
  appTitle: "Notes",
  noteLoadFailure: "Cannot load note",
  notesLoadFailure: "Cannot load notes",

  noteCreateSuccess: "New note added",
  noteCreateFailure: "Cannot add new note",
  noteUpdateSuccess: "Note updated",
  noteUpdateFailure: "Cannot update note",
  noteDeleteSuccess: "Note deleted",
  noteDeleteFailure: "Cannot delete note",

  formTitle: "Title",
  formSave: "Save",
  formDelete: "Delete",

  editFormHeader: "Note details",
  newFormHeader: "Create new note",
}

const czPhrases: Phrases = {
  appTitle: "Poznámky",
  noteLoadFailure: "Nelze načíst poznámku",
  notesLoadFailure: "Nelze načíst poznámky",

  noteCreateSuccess: "Přidána poznámka",
  noteCreateFailure: "Nelze přidat poznámku",
  noteUpdateSuccess: "Poznámka upravena",
  noteUpdateFailure: "Nelze upravit poznámku",
  noteDeleteSuccess: "Poznámka smazána",
  noteDeleteFailure: "Nelze smazat poznámku",

  formTitle: "Nadpis",
  formSave: "Uložit",
  formDelete: "Smazat",

  editFormHeader: "Detaily poznámky",
  newFormHeader: "Vytvořit novou poznámku",
}

export const getPhrase = (language: "cz" | "en", key: PhraseKeys): string => {
  switch (language) {
    case "cz":
      return czPhrases[key]
    case "en":
      return enPhrases[key]
  }
}

export const useTranslate = () => {
  const language = useSelector((state: RootState) => state.language.code)
  return (key: PhraseKeys) => getPhrase(language, key)
}
