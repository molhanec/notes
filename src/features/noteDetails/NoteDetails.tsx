import React, { useEffect } from "react"
import { connect, ResolveThunks } from "react-redux"

import { RootState } from "../../app/store"
import { loadNoteDetailsAsync, setNoteDetails } from "./noteDetailsSlice"
import { RouteComponentProps } from "react-router-dom"
import EditNoteForm from "../noteForm/EditNoteForm"
import { deleteNoteAsync } from "../../app/data"

type RouterParams = RouteComponentProps<{
  noteId: string
}>

type Props = ReturnType<typeof mapStateToProps> &
  ResolveThunks<typeof mapDispatchToProps> &
  RouterParams

export const NoteDetails: React.FC<Props> = ({
  match,
  noteDetails,
  loadNoteDetailsAsync,
  setNoteDetails,
  deleteNoteAsync,
}: Props) => {
  useEffect(() => {
    setNoteDetails(undefined)
    loadNoteDetailsAsync(match.params.noteId)
  }, [loadNoteDetailsAsync, setNoteDetails, match.params.noteId])

  if (!noteDetails.id) return null

  const handleDelete = () => {
    if (!noteDetails.id) return
    deleteNoteAsync(noteDetails.id)
  }

  return (
    <>
      <EditNoteForm onDelete={handleDelete} />
    </>
  )
}

const mapStateToProps = (state: RootState) => ({
  noteDetails: state.noteDetails,
})

const mapDispatchToProps = {
  loadNoteDetailsAsync,
  setNoteDetails,
  deleteNoteAsync,
}

const ConnectedNoteDetails = connect(
  mapStateToProps,
  mapDispatchToProps
)(NoteDetails)
export default ConnectedNoteDetails
