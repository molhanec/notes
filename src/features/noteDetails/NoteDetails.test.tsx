import React from "react"
import { render, fireEvent } from "@testing-library/react"
import { Provider } from "react-redux"
import { store } from "../../app/store"
import { NoteDetails } from "./NoteDetails"
import { Router } from "react-router-dom"
import { createMemoryHistory } from "history"
import { getPhrase } from "../../app/translations"

describe("<NoteDetails>", () => {
  it("matches snapshot", () => {
    const loadNoteDetailsAsync = jest.fn()
    const setNoteDetails = jest.fn()
    const deleteNoteAsync = jest.fn()
    const { container } = render(
      <Provider store={store}>
        <Router history={createMemoryHistory()}>
          <NoteDetails
            noteDetails={{
              id: 1,
              title: "title 1",
            }}
            match={{
              params: {
                noteId: "1",
              },
              isExact: false,
              path: "",
              url: "",
            }}
            loadNoteDetailsAsync={loadNoteDetailsAsync}
            setNoteDetails={setNoteDetails as any}
            deleteNoteAsync={deleteNoteAsync}
            history={{} as any}
            location={{} as any}
          />
        </Router>
      </Provider>
    )

    expect(container).toMatchSnapshot()
    expect(loadNoteDetailsAsync).toBeCalled()
    expect(setNoteDetails).toBeCalledWith(undefined)
  })

  it("calls deleteNoteAsync when Delete button clicked", () => {
    const loadNoteDetailsAsync = jest.fn()
    const setNoteDetails = jest.fn()
    const deleteNoteAsync = jest.fn()
    const { getByText } = render(
      <Provider store={store}>
        <Router history={createMemoryHistory()}>
          <NoteDetails
            noteDetails={{
              id: 1,
              title: "title 1",
            }}
            match={{
              params: {
                noteId: "1",
              },
              isExact: false,
              path: "",
              url: "",
            }}
            loadNoteDetailsAsync={loadNoteDetailsAsync}
            setNoteDetails={setNoteDetails as any}
            deleteNoteAsync={deleteNoteAsync}
            history={{} as any}
            location={{} as any}
          />
        </Router>
      </Provider>
    )

    const deleteButton = getByText(getPhrase("cz", "formDelete"))
    fireEvent.click(deleteButton)

    expect(deleteNoteAsync).toBeCalled()
  })
})
