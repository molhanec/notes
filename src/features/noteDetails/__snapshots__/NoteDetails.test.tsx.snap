// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteDetails> matches snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Detaily poznámky
    </div>
    <div
      class="card-body"
    >
      <form
        class=""
      >
        <label
          class=""
          for="title"
        >
          Nadpis
        </label>
        <input
          class="form-control"
          id="title"
          type="text"
          value=""
        />
        <button
          class="mt-3 btn btn-primary disabled"
          disabled=""
        >
          Uložit
        </button>
        <button
          class="mt-3 ml-1 btn btn-danger"
          type="button"
        >
          Smazat
        </button>
      </form>
    </div>
  </div>
</div>
`;
